File: backend/agent/tools/files.py
from __future__ import annotations

import asyncio
import os
import pathlib
import re
import shutil
import subprocess
from pathlib import Path
from typing import Any

from agent.tools.base import Tool, ToolResult, _diff_full, _diff_stats, _in_cwd


class FileReadTool(Tool):
    name = "file_read"
    description = (
        "Read a file with line numbers. "
        "offset: first line to read (1-based). limit: max lines to return. "
        "Omit both to read the whole file."
    )
    params = {
        "path": {"type": "string", "description": "File path"},
        "offset": {"type": "integer", "description": "First line (1-based, optional)"},
        "limit": {"type": "integer", "description": "Max lines to return (optional)"},
    }

    def tool_definition(self) -> dict:
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": self.params,
                    "required": ["path"],
                },
            },
        }

    async def execute(self, args: dict[str, Any]) -> str:
        path = args.get("path", "")
        offset = int(args.get("offset") or 1)
        limit = args.get("limit")
        if offset < 1:
            offset = 1
        try:
            try:
                if os.path.getsize(path) > 50 * 1024 * 1024:
                    return f"error: file too large (>50MB) — use offset/limit or grep: {path}"
            except OSError:
                pass
            with open(path, encoding="utf-8", errors="replace") as f:
                raw_lines = f.readlines()
            start = offset - 1
            end = start + int(limit) if limit else len(raw_lines)
            slice_lines = raw_lines[start:end]
            result = []
            for i, line in enumerate(slice_lines, start + 1):
                line = line.rstrip("\n")
                if len(line) > 2000:
                    line = line[:2000] + "…"
                result.append(f"{i}: {line}")
            content = "\n".join(result)
            if len(content) > 8000:
                return content[:8000] + "\n...(truncated)"
            return content
        except Exception as e:
            return f"error: {e}"


class GlobTool(Tool):
    name = "glob"
    description = (
        "Find files by glob pattern. Returns matching paths sorted by modification time. "
        "Use ** for recursive search, e.g. **/*.py"
    )
    params = {
        "pattern": {"type": "string", "description": "Glob pattern, e.g. **/*.py"},
        "path": {"type": "string", "description": "Base directory (default: current)"},
    }

    def tool_definition(self) -> dict:
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": self.params,
                    "required": ["pattern"],
                },
            },
        }

    async def execute(self, args: dict[str, Any]) -> str:
        pattern = args.get("pattern", "")
        base = pathlib.Path(args.get("path") or ".").resolve()

        def _mtime(p: pathlib.Path) -> float:
            try:
                return p.stat().st_mtime
            except OSError:
                return 0.0  # broken symlink / vanished file — sort last, don't crash

        try:
            matches = sorted(base.glob(pattern), key=_mtime, reverse=True)
            if not matches:
                return "(no matches)"
            paths = [str(p.relative_to(base)).replace("\\", "/") for p in matches[:200]]
            result = "\n".join(paths)
            if len(matches) > 200:
                result += f"\n... and {len(matches) - 200} more"
            return result
        except Exception as e:
            return f"error: {e}"


class GrepTool(Tool):
    name = "grep"
    description = (
        "Search file contents by regex. Returns file:line: text for each match. "
        "Uses ripgrep if available, falls back to Python."
    )
    params = {
        "pattern": {"type": "string", "description": "Regex pattern"},
        "path": {"type": "string", "description": "Directory or file to search (default: .)"},
        "glob": {"type": "string", "description": "File filter, e.g. *.py (optional)"},
        "case_insensitive": {"type": "boolean", "description": "Ignore case (default false)"},
    }

    def tool_definition(self) -> dict:
        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": self.params,
                    "required": ["pattern"],
                },
            },
        }

    async def execute(self, args: dict[str, Any]) -> str:
        pattern = args.get("pattern", "")
        path = args.get("path") or "."
        glob_filter = args.get("glob", "")
        nocase = args.get("case_insensitive", False)

        rg = shutil.which("rg")
        if rg:
            cmd = [rg, "--line-number", "--no-heading", "--color=never", "--max-count=100"]
            if nocase:
                cmd.append("-i")
            if glob_filter:
                cmd += ["--glob", glob_filter]
            cmd += [pattern, path]
            try:
                proc = await asyncio.to_thread(
                    subprocess.run,
                    cmd,
                    capture_output=True,
                    text=True,
                    timeout=15,
                    encoding="utf-8",
                    errors="replace",
                )
                out = proc.stdout.strip()
                if not out:
                    return "(no matches)"
                lines = out.split("\n")
                if len(lines) > 100:
                    return "\n".join(lines[:100]) + f"\n... and {len(lines) - 100} more"
                return out
            except Exception:
                pass

        # Python fallback
        try:
            flags = re.IGNORECASE if nocase else 0
            regex = re.compile(pattern, flags)
            base = pathlib.Path(path)
            if base.is_file():
                files = [base]
            else:
                files = list(base.rglob(glob_filter or "*"))
            results = []
            for f in sorted(files):
                if not f.is_file():
                    continue
                try:
                    for i, line in enumerate(
                        f.read_text(encoding="utf-8", errors="ignore").splitlines(), 1
                    ):
                        if regex.search(line):
                            rel = str(f.relative_to(base) if not base.is_file() else f).replace(
                                "\\", "/"
                            )
                            results.append(f"{rel}:{i}: {line.rstrip()}")
                            if len(results) >= 100:
                                break
                except Exception:
                    pass
                if len(results) >= 100:
                    break
            return "\n".join(results) if results else "(no matches)"
        except Exception as e:
            return f"error: {e}"


class EditTool(Tool):
    name = "edit"
    description = (
        "Replace a specific string in a file. "
        "old_string must match exactly once — add surrounding lines to make it unique. "
        "For new files use file_write instead."
    )
    params = {
        "path": {"type": "string", "description": "File path"},
        "old_string": {
            "type": "string",
            "description": "Exact text to replace (must be unique in the file)",
        },
        "new_string": {"type": "string", "description": "Replacement text"},
    }

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        return not _in_cwd(args.get("path", ""))

    async def execute(self, args: dict[str, Any]) -> str:
        path = args.get("path", "")
        old = args.get("old_string", "")
        new = args.get("new_string", "")
        if not old:
            return "error: old_string is empty"
        try:
            with open(path, encoding="utf-8") as f:
                content = f.read()
            count = content.count(old)
            if count == 0:
                # Try with normalised line endings
                norm_content = content.replace("\r\n", "\n")
                norm_old = old.replace("\r\n", "\n")
                if norm_content.count(norm_old) == 1:
                    new_content = norm_content.replace(norm_old, new.replace("\r\n", "\n"), 1)
                    with open(path, "w", encoding="utf-8") as f:
                        f.write(new_content)
                    stats = _diff_stats(old, new)
                    label = f"edited {path}  {stats}" if stats else f"edited {path}"
                    return ToolResult(label, _diff_full(old, new, path))
                return f"error: old_string not found in {path}"
            if count > 1:
                return f"error: old_string matches {count} times — make it more specific"
            new_content = content.replace(old, new, 1)
            with open(path, "w", encoding="utf-8") as f:
                f.write(new_content)
            stats = _diff_stats(old, new)
            label = f"edited {path}  {stats}" if stats else f"edited {path}"
            return ToolResult(label, _diff_full(old, new, path))
        except FileNotFoundError:
            return f"error: file not found: {path}"
        except Exception as e:
            return f"error: {e}"


class FileWriteTool(Tool):
    name = "file_write"
    description = "Write text to a file. Creates directories if needed."
    params = {
        "path": {"type": "string", "description": "File path"},
        "content": {"type": "string", "description": "Content"},
    }

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        return not _in_cwd(args.get("path", ""))

    async def execute(self, args: dict[str, Any]) -> str:
        path = args.get("path", "")
        content = args.get("content", "")
        try:
            d = os.path.dirname(os.path.abspath(path))
            if d:
                os.makedirs(d, exist_ok=True)
            old_content = ""
            try:
                with open(path, encoding="utf-8") as f:
                    old_content = f.read()
            except (FileNotFoundError, OSError):
                pass
            with open(path, "w", encoding="utf-8") as f:
                f.write(content)
            stats = _diff_stats(old_content, content)
            label = f"wrote {path}  {stats}" if stats else f"wrote {path}"
            diff = _diff_full(old_content, content, path) if stats else None
            return ToolResult(label, diff)
        except Exception as e:
            return f"error: {e}"


class FileDeleteTool(Tool):
    name = "file_delete"
    description = "Delete a file. Only works inside the current directory."
    params = {"path": {"type": "string", "description": "File path"}}

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        return True

    async def execute(self, args: dict[str, Any]) -> str:
        path = args.get("path", "")
        try:
            abs_path = os.path.abspath(path)
            if not _in_cwd(path):
                return "error: cannot delete files outside current directory"
            if not os.path.exists(abs_path):
                return f"error: not found: {path}"
            if os.path.isdir(abs_path):
                return "error: this is a directory — use bash rmdir"
            os.remove(abs_path)
            return f"deleted: {path}"
        except Exception as e:
            return f"error: {e}"


class FileListTool(Tool):
    name = "file_list"
    description = "List files and directories."
    params = {"path": {"type": "string", "description": "Directory (. for current)"}}

    async def execute(self, args: dict[str, Any]) -> str:
        path = args.get("path", ".")
        try:
            from agent.safety._filesystem import is_path_within_workspace

            target = Path(path).resolve()
            workspace = Path.cwd().resolve()
            if not is_path_within_workspace(target, workspace):
                return "error: path is outside allowed scope"
            with os.scandir(path) as entries_iter:
                entries = sorted(entries_iter, key=lambda e: e.name.lower())
            dirs, files = [], []
            for e in entries:
                # Skip symlinks that escape the workspace
                if e.is_symlink():
                    resolved = Path(e.path).resolve()
                    if not is_path_within_workspace(resolved, workspace):
                        continue
                if e.is_dir():
                    dirs.append(f"[{e.name}]")
                else:
                    size = e.stat().st_size
                    kb = size / 1024
                    files.append(f"{e.name} ({kb:.1f}KB)" if kb >= 1 else f"{e.name} ({size}B)")
            parts = []
            if dirs:
                parts.append("dirs: " + "  ".join(dirs))
            if files:
                parts.append("files: " + "  ".join(files))
            return "\n".join(parts) if parts else "(empty)"
        except Exception as e:
            return f"error: {e}"

File: agent/tools/files.go
package tools

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// --- FileReadTool ---

type FileReadTool struct{}

func (t *FileReadTool) Name() string { return "file_read" }

func (t *FileReadTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Read a UTF-8 text file from the workspace.",
		Params: map[string]any{
			"path":   prop("string", "Absolute or relative path to read"),
			"offset": prop("integer", "Line number to start from (1-based)"),
			"limit":  prop("integer", "Maximum number of lines to return"),
		},
		Required: []string{"path"},
	}
}

func (t *FileReadTool) Execute(args map[string]any) Result {
	path := getStr(args, "path")
	offset := getInt(args, "offset", 1)
	if offset < 1 {
		offset = 1
	}
	limit := getInt(args, "limit", 0)

	fi, err := os.Stat(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if fi.IsDir() {
		return R("error: is a directory, not a file: " + path)
	}
	if fi.Size() > 50*1024*1024 {
		return R("error: file too large (>50MB) — use offset/limit or grep: " + path)
	}

	f, err := os.Open(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer f.Close()

	var rawLines []string
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
	for scanner.Scan() {
		rawLines = append(rawLines, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return R("error: " + err.Error())
	}

	start := offset - 1
	end := len(rawLines)
	if limit > 0 {
		end = start + limit
	}
	if start >= len(rawLines) {
		return R("")
	}
	if end > len(rawLines) {
		end = len(rawLines)
	}
	slice := rawLines[start:end]

	var b strings.Builder
	for i, line := range slice {
		lineNum := start + i + 1
		if len(line) > 2000 {
			line = line[:2000] + "…"
		}
		b.WriteString(fmt.Sprintf("%d: %s\n", lineNum, line))
	}
	result := b.String()
	if len(result) > 8000 {
		result = result[:8000] + "\n...(truncated)"
	}
	return R(result)
}

// --- GlobTool ---

type GlobTool struct{}

func (t *GlobTool) Name() string { return "glob" }

func (t *GlobTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Find files matching a glob pattern, newest first.",
		Params: map[string]any{
			"pattern": prop("string", "Glob pattern, e.g. src/**/*.ts"),
			"path":    prop("string", "Root directory to search from (default: cwd)"),
		},
		Required: []string{"pattern"},
	}
}

func (t *GlobTool) Execute(args map[string]any) Result {
	pattern := getStr(args, "pattern")
	basePath := getStr(args, "path")
	if basePath == "" {
		basePath = "."
	}
	base, err := filepath.Abs(basePath)
	if err != nil {
		return R("error: " + err.Error())
	}

	// Fast path: use filepath.Glob if no **, otherwise walk
	var matches []string
	if strings.Contains(pattern, "**") {
		err = filepath.Walk(base, func(p string, fi os.FileInfo, err error) error {
			if err != nil {
				return nil
			}
			matched, err := filepath.Match(pattern, p)
			if err != nil {
				return nil
			}
			// Also try matching relative to base
			if !matched {
				rel, _ := filepath.Rel(base, p)
				matched, _ = filepath.Match(pattern, rel)
				if !matched {
					matched, _ = filepath.Match(pattern, strings.ReplaceAll(rel, "\\", "/"))
				}
			}
			if matched {
				matches = append(matches, p)
			}
			return nil
		})
	} else {
		matches, err = filepath.Glob(filepath.Join(base, pattern))
	}
	if err != nil {
		return R("error: " + err.Error())
	}

	// Sort by modification time, newest first
	sort.Slice(matches, func(i, j int) bool {
		fi, err1 := os.Stat(matches[i])
		fj, err2 := os.Stat(matches[j])
		if err1 != nil || err2 != nil {
			return matches[i] < matches[j]
		}
		return fi.ModTime().After(fj.ModTime())
	})

	if len(matches) == 0 {
		return R("(no matches)")
	}

	var b strings.Builder
	count := 0
	for _, p := range matches {
		if count >= 200 {
			b.WriteString(fmt.Sprintf("\n... and %d more", len(matches)-200))
			break
		}
		rel, _ := filepath.Rel(base, p)
		rel = strings.ReplaceAll(rel, "\\", "/")
		if count > 0 {
			b.WriteString("\n")
		}
		b.WriteString(rel)
		count++
	}
	return R(b.String())
}

// --- GrepTool ---

type GrepTool struct{}

func (t *GrepTool) Name() string { return "grep" }

func (t *GrepTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Search file contents with a regular expression.",
		Params: map[string]any{
			"pattern":          prop("string", "Regular expression to search"),
			"path":             prop("string", "File or directory to search in (default: cwd)"),
			"glob":             prop("string", "Only search files whose name matches this glob"),
			"case_insensitive": prop("boolean", "Case-insensitive match"),
		},
		Required: []string{"pattern"},
	}
}

func (t *GrepTool) Execute(args map[string]any) Result {
	pattern := getStr(args, "pattern")
	path := getStr(args, "path")
	if path == "" {
		path = "."
	}
	globFilter := getStr(args, "glob")
	nocase, _ := args["case_insensitive"].(bool)

	// Try rg first
	if rgPath, err := exec.LookPath("rg"); err == nil {
		cmdArgs := []string{"--line-number", "--no-heading", "--color=never", "--max-count=100"}
		if nocase {
			cmdArgs = append(cmdArgs, "-i")
		}
		if globFilter != "" {
			cmdArgs = append(cmdArgs, "--glob", globFilter)
		}
		cmdArgs = append(cmdArgs, pattern, path)
		cmd := exec.Command(rgPath, cmdArgs...)
		out, err := cmd.Output()
		if err == nil {
			output := strings.TrimSpace(string(out))
			if output == "" {
				return R("(no matches)")
			}
			lines := strings.Split(output, "\n")
			if len(lines) > 100 {
				return R(strings.Join(lines[:100], "\n") + fmt.Sprintf("\n... and %d more", len(lines)-100))
			}
			return R(output)
		}
	}

	// Go fallback
	rePattern := pattern
	if nocase {
		rePattern = "(?i)" + pattern
	}
	re, err := regexp.Compile(rePattern)
	if err != nil {
		return R("error: " + err.Error())
	}

	base, err := filepath.Abs(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	fi, err := os.Stat(base)
	if err != nil {
		return R("error: " + err.Error())
	}

	var files []string
	if !fi.IsDir() {
		files = []string{base}
	} else {
		walkFn := func(p string, info os.FileInfo, err error) error {
			if err != nil || info.IsDir() {
				return nil
			}
			if globFilter != "" {
				matched, _ := filepath.Match(globFilter, info.Name())
				if !matched {
					return nil
				}
			}
			files = append(files, p)
			return nil
		}
		filepath.Walk(base, walkFn)
	}
	sort.Strings(files)

	var results []string
	for _, f := range files {
		fi, err := os.Stat(f)
		if err != nil || fi.IsDir() || fi.Size() > 10*1024*1024 {
			continue
		}
		data, err := os.ReadFile(f)
		if err != nil {
			continue
		}
		rel, _ := filepath.Rel(base, f)
		if fi.IsDir() {
			rel = f
		}
		rel = strings.ReplaceAll(rel, "\\", "/")
		for i, line := range strings.Split(string(data), "\n") {
			if re.MatchString(line) {
				results = append(results, fmt.Sprintf("%s:%d: %s", rel, i+1, strings.TrimRight(line, "\r")))
				if len(results) >= 100 {
					break
				}
			}
		}
		if len(results) >= 100 {
			break
		}
	}
	if len(results) == 0 {
		return R("(no matches)")
	}
	return R(strings.Join(results, "\n"))
}

// --- EditTool ---

type EditTool struct{}

func (t *EditTool) Name() string { return "edit" }

func (t *EditTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Apply a precise string replacement to a file. old_string must occur exactly once.",
		Params: map[string]any{
			"path":       prop("string", "File to edit"),
			"old_string": prop("string", "Exact string to replace"),
			"new_string": prop("string", "Replacement string"),
		},
		Required: []string{"path", "old_string", "new_string"},
	}
}

func (t *EditTool) Execute(args map[string]any) Result {
	path := getStr(args, "path")
	oldStr := getStr(args, "old_string")
	newStr := getStr(args, "new_string")
	if oldStr == "" {
		return R("error: old_string is empty")
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	content := string(data)
	count := strings.Count(content, oldStr)

	if count == 0 {
		normContent := strings.ReplaceAll(content, "\r\n", "\n")
		normOld := strings.ReplaceAll(oldStr, "\r\n", "\n")
		if strings.Count(normContent, normOld) == 1 {
			newContent := strings.Replace(normContent, normOld, strings.ReplaceAll(newStr, "\r\n", "\n"), 1)
			if err := os.WriteFile(path, []byte(newContent), 0644); err != nil {
				return R("error: " + err.Error())
			}
			stats := diffStats(oldStr, newStr)
			label := fmt.Sprintf("edited %s", path)
			if stats != "" {
				label += "  " + stats
			}
			return RD(label, diffFull(oldStr, newStr, path))
		}
		return R("error: old_string not found in " + path)
	}
	if count > 1 {
		return R(fmt.Sprintf("error: old_string matches %d times — make it more specific", count))
	}
	newContent := strings.Replace(content, oldStr, newStr, 1)
	if err := os.WriteFile(path, []byte(newContent), 0644); err != nil {
		return R("error: " + err.Error())
	}
	stats := diffStats(oldStr, newStr)
	label := fmt.Sprintf("edited %s", path)
	if stats != "" {
		label += "  " + stats
	}
	return RD(label, diffFull(oldStr, newStr, path))
}

// --- FileWriteTool ---

type FileWriteTool struct{}

func (t *FileWriteTool) Name() string { return "file_write" }

func (t *FileWriteTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Write text content to a file inside the workspace, creating parent dirs.",
		Params: map[string]any{
			"path":    prop("string", "Path to write to"),
			"content": prop("string", "Text content to write"),
		},
		Required: []string{"path", "content"},
	}
}

func (t *FileWriteTool) Execute(args map[string]any) Result {
	path := getStr(args, "path")
	content := getStr(args, "content")
	absPath, err := filepath.Abs(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if d := filepath.Dir(absPath); d != "" {
		os.MkdirAll(d, 0755)
	}
	var oldContent string
	if data, err := os.ReadFile(path); err == nil {
		oldContent = string(data)
	}
	if err := os.WriteFile(path, []byte(content), 0644); err != nil {
		return R("error: " + err.Error())
	}
	stats := diffStats(oldContent, content)
	label := fmt.Sprintf("wrote %s", path)
	if stats != "" {
		label += "  " + stats
		return RD(label, diffFull(oldContent, content, path))
	}
	return R(label)
}

// --- FileDeleteTool ---

type FileDeleteTool struct{}

func (t *FileDeleteTool) Name() string { return "file_delete" }

func (t *FileDeleteTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "Delete a file or directory inside the workspace.",
		Params: map[string]any{
			"path":      prop("string", "Path to delete"),
			"recursive": prop("boolean", "Required to delete a non-empty directory"),
		},
		Required: []string{"path"},
	}
}

func (t *FileDeleteTool) Execute(args map[string]any) Result {
	path := getStr(args, "path")
	recursive, _ := args["recursive"].(bool)
	absPath, err := filepath.Abs(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if !inCwd(path) {
		return R("error: cannot delete outside current directory")
	}
	if _, err := os.Stat(absPath); os.IsNotExist(err) {
		return R("error: not found: " + path)
	}
	fi, err := os.Stat(absPath)
	if err != nil {
		return R("error: " + err.Error())
	}
	if fi.IsDir() {
		if !recursive {
			return R("error: is a directory — set recursive=true to delete")
		}
		os.RemoveAll(absPath)
		return R("deleted directory: " + path)
	}
	os.Remove(absPath)
	return R("deleted: " + path)
}

// --- FileListTool ---

type FileListTool struct{}

func (t *FileListTool) Name() string { return "file_list" }

func (t *FileListTool) Spec() ToolSpec {
	return ToolSpec{
		Description: "List entries of a directory.",
		Params: map[string]any{
			"path": prop("string", "Directory to list (default: cwd)"),
		},
	}
}

func (t *FileListTool) Execute(args map[string]any) Result {
	path := getStr(args, "path")
	if path == "" {
		path = "."
	}
	abs, err := filepath.Abs(path)
	if err != nil {
		return R("error: " + err.Error())
	}
	if !inCwd(path) {
		return R("error: path is outside allowed scope")
	}

	d, err := os.Open(abs)
	if err != nil {
		return R("error: " + err.Error())
	}
	defer d.Close()

	entries, err := d.Readdir(-1)
	if err != nil {
		return R("error: " + err.Error())
	}
	sort.Slice(entries, func(i, j int) bool {
		return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
	})

	ws, _ := filepath.Abs(".")
	var dirs, files []string
	for _, e := range entries {
		name := e.Name()
		if e.Mode()&os.ModeSymlink != 0 {
			resolved, err := filepath.EvalSymlinks(filepath.Join(abs, name))
			if err != nil {
				continue
			}
			rel, err := filepath.Rel(ws, resolved)
			if err != nil || strings.HasPrefix(rel, "..") {
				continue
			}
		}
		if e.IsDir() {
			dirs = append(dirs, "["+name+"]")
		} else {
			size := e.Size()
			if kb := float64(size) / 1024; kb >= 1 {
				files = append(files, fmt.Sprintf("%s (%.1fKB)", name, kb))
			} else {
				files = append(files, fmt.Sprintf("%s (%dB)", name, size))
			}
		}
	}
	var b strings.Builder
	if len(dirs) > 0 {
		b.WriteString("dirs: " + strings.Join(dirs, "  "))
	}
	if len(files) > 0 {
		if b.Len() > 0 {
			b.WriteString("\n")
		}
		b.WriteString("files: " + strings.Join(files, "  "))
	}
	if b.Len() == 0 {
		return R("(empty)")
	}
	return R(b.String())
}

func execLookPathFn(file string) (string, error) {
	return exec.LookPath(file)
}

func execCommandFn(name string, arg ...string) *exec.Cmd {
	return exec.Command(name, arg...)
}
